### ISSUE TITLE
Support for AWS CDK

### ISSUE THREAD

**Issue by moravcik** [serverless/serverless#5573]
# This is a Feature Proposal

## Description

Are there any plans to support or integrate into AWS CDK (Cloud Development Kit)?
https://awslabs.github.io/aws-cdk/

**AWS CDK** can nicely handle CloudFormation Stack creation and updates using TypeScript, but lacks packaging lambda functions for deployment. For deploying lambda the AWS CDK supports only inline code, S3 archive and local directory. 

**Serverless** is managing its own CloudFormation stack with custom resources, and handles packaging of lambda nicely. 

How about to integrate these two tools?

There is one similar question on forum (yet unanswered):
https://forum.serverless.com/t/aws-cdk-vs-serverless-framework-for-lambda/5673

**Comment by asyba** [comment]
Yeah, and also I would like to create serverless template through code.

**Comment by bradennapier** [comment]
Mixing the two seems like an awesome combination if done right!

**Comment by karfau** [comment]
I wonder what is missing for this to work.
(I only started diving into this so sorry in case I ask soething that should be obvious.)
As of writing this the [CDK allows to include existing CloudFormation config into a stack](https://awslabs.github.io/aws-cdk/refs/_aws-cdk_aws-s3.html#importing-existing-buckets):
and I understood that it is possible to get the resulting CloudFormation config when running `sls package`.
I tried to import the following things:
- `cloudformation-template-create-stack.json` + `cloudformation-template-update-stack.json`
- only `cloudformation-template-create-stack.json`
- only `cloudformation-template-update-stack.json`
- `serverless-state.json` (the path `.service.provider.compiledCloudFormationTemplate`)

But when CDK tries to create the config it has some Error in some `merge` function:
```
/.../node_modules/@aws-cdk/cdk/lib/stack.ts:580
  for (const section of Object.keys(part)) {
                               ^
TypeError: Cannot convert undefined or null to object
```

Is this supposed to work and it might just be my config (I don't ask for iscussing those here, since it's out of the scope of this issue)?

**Comment by maxsap** [comment]
+1 for this feature

**Comment by dgvivero** [comment]
+1

**Comment by dav2490** [comment]
+1

**Comment by Adrxx** [comment]
+1

**Comment by BerkhanBerkdemir** [comment]
*Not cool* to spam with +1 **comments**.

I am very long time user of Serverless and very grateful about what Serverless framework brings into community. Adding AWS CDK **would be** great; however, just to mention that may be broken the current services because currently we statically added these CloudFormation templates.

**Comment by jalchr** [comment]
How is what you are requesting is different from this ?
https://docs.aws.amazon.com/cdk/latest/guide/serverless_example.html

May be this is an old issue

**Comment by niklaswallerstedt** [comment]
I really like both Serverless and CDK, so any tips on how to do combine the two would be appreciated. Coming from Serverless and just recently authoring lambda functions in CDK this is my personal take:

- Missing serverless-offline for local debugging and other plugins. After some research I can run the functions with SAM but that requires me to copy the logical id.
https://docs.aws.amazon.com/cdk/latest/guide/tools.html#sam

`sam local invoke {{LogicalID}}`

https://github.com/aws/aws-cdk/issues/716
> Closing this issue since we don't have anything concrete to do here at this time. In general, better interplay between the CDK and SAM would be to an advantage. However, we don't have anything to offer just yet.

- Adding third party libraries has to be done to each lambda individually, converted my CDK-project to a yarn workspace to make this easier to manage. Currently investigating adding webpack to handle packaging/bundling better.

On the plus side for CDK, the ease of provisioning infrastructure, granting permissions to lambda functions, etc is great.

```
const table = new dynamodb.Table(this, "Table", {
      partitionKey: {
        name: "id",
        type: dynamodb.AttributeType.STRING
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    const fn = new lambda.Function(this, "Lambda", {
      runtime: lambda.Runtime.NODEJS_12_X,
      code: lambda.Code.fromAsset("./lambda"),
      handler: "my-function.handler"
    });

    fn.addEnvironment("DYNAMODB_TABLE", table.tableName);
    table.grantWriteData(fn);
```

**Comment by danilofuchs** [comment]
I have an idea of how they could both work together, but I didn't try it yet:
Create a CDK 'infra' stack with the most 'permanent' resources, such as Dynamo tables, SQS queues, etc.

In the Serverless config, somehow state that the 'app' (with the volatile functions) stack depends on the infra stack.

In the CI environment, always update infra before app stack

**Comment by RichardDRJ** [comment]
We've been using a custom plugin for Serverless Framework for a wee while which allows us to specify most of the infrastructure around Lambdas with CDK, while still getting some benefits provided by Serverless, such as packaging. We've just published it publicly in the hope that this scratches the itch for some other folks.

Right now it's primarily focussed around wrapping core CDK functionality up in an SLS plugin, so it supports `deploy`, `remove`, and `diff`. The hope is to enable things like `invoke` in future.

Please feel free to raise issues or pull requests if you can see a way to make it better!

GH Repo: https://github.com/snap40/serverless-aws-cdk
NPM package: https://www.npmjs.com/package/serverless-aws-cdk

**Comment by danilofuchs** [comment]
> We've been using a custom plugin for Serverless Framework for a wee while which allows us to specify most of the infrastructure around Lambdas with CDK, while still getting some benefits provided by Serverless, such as packaging. We've just published it publicly in the hope that this scratches the itch for some other folks.
> 
> Right now it's primarily focussed around wrapping core CDK functionality up in an SLS plugin, so it supports `deploy`, `remove`, and `diff`. The hope is to enable things like `invoke` in future.
> 
> Please feel free to raise issues or pull requests if you can see a way to make it better!
> 
> GH Repo: https://github.com/snap40/serverless-aws-cdk
> NPM package: https://www.npmjs.com/package/serverless-aws-cdk

Thanks! It looks very promising, I will try it out soon. Using both serverless and CDK on a daily basis, combining the two seems to be the optimal path.

**Comment by fwang** [comment]
We actually put something together recently at Serverless Stack to helping using Serverless Framework and AWS CDK together. Serverless’s strength is working with Lambda and API Gateway. So a good pattern here is to use CDK for the rest of your infrastructure, and Serverless for your Lambda functions.

So we created a tool to do exactly this - Serverless Stack Toolkit (SST), which allows you to combine CDK and Serverless Framework - https://github.com/serverless-stack/serverless-stack

You can deploy your Serverless service and CDK service like `sls deploy --stage dev` and `sst deploy --stage dev`.

**Comment by Birowsky** [comment]
@fwang does this mean that lambdas keep being described with yaml?

**Comment by fwang** [comment]
The functions section in serverless.yml is pretty concise. So yeah, we still use Serverless Framework to define, build/webpack and deploy our Lambdas.

We always deploy the SST (resources) service first, and then deploy the Serverless service.

**Comment by mnapoli** [comment]
Hey everyone, we'd love to collect a bit more info on what you would do with CDK within/with Framework? **What would be your use case?** (the more specific you are the better)

What I've noted so far is around "shared resources" (aka "infra stack"), i.e. I'm guessing to avoid writing CloudFormation in `serverless.yml`? Any other use case? What kind of resources would you deploy? Are we talking about Lambda functions, or other resources like databases, queues, buckets, etc.

**Comment by alfaproject** [comment]
We currently create some of our CF resources using the CDK exactly as you mentioned. We have SQS, Kinesis, CW metrics, SNS etc

**Comment by mnapoli** [comment]
@coyoteecd we are exploring ideas for projects that contain multiple stacks/multiple services, it seems to be your case? If it is, I'd love to learn more about your setup and get your feedback on an idea we are exploring. Feel free to send me an email (matthieu.napoli at serverless dot com) or a direct message in the "Serverless community" slack ([invite link](https://join.slack.com/t/serverless-contrib/shared_invite/zt-13unxiht3-wTSFk910WilkYNpECmlPDA)).

That goes for anyone else reading this btw, feel free to get in touch with me (@alfaproject I sent you an email too). Currently we are trying to learn more how everyone deploys multiple stacks. Once we get a better picture, we'll post some ideas in an issue in this repository.

**Comment by automartin5000** [comment]
It’s not just for other resources, although that’s important too. Serverless itself could have a CDK construct. Take a look at the Chalice CDK integration. It’s pretty cool. There’s a `Chalice` [CDK construct](https://aws.amazon.com/blogs/developer/aws-chalice-adds-support-for-the-aws-cdk/). It allows you to deploy your whole stack with just `cdk deploy`.

**Comment by mnapoli** [comment]
@automartin5000 thanks for the link, I had no idea Chalice had a CDK construct 🤔

**Comment by automartin5000** [comment]
> @automartin5000 thanks for the link, I had no idea Chalice had a CDK construct 🤔

So does a “Serverless” CDK construct seem like something the Serverless team would take on?

**Comment by mnapoli** [comment]
Here's a quick update:

2 weeks ago, we released [Serverless Framework Compose](https://www.serverless.com/blog/serverless-framework-compose-multi-service-deployments), which focuses on multi-service deployments (as mentioned above in my previous comment).

Compose orchestrates multiple Serverless Framework services. Now we are exploring introducing other kinds of "services" in `serverless-compose.yml`. For example deploying CDK stacks/constructs alongside Serverless Framework services (via `serverless-compose.yml`).

Implementation-wise, this is not a small task, so we are looking at it step by step. This is why it's a bit early to say exactly what we will do, and how others can help. But if you have specific examples of what you would deploy with this feature, feel free to share them here. Examples are always welcome, they help us project our ideas to real scenarios.
